### ISSUE TITLE
adapter-node's handler should never show a 404

### ISSUE THREAD

**Issue by GauBen** [sveltejs/kit#3363]
### Describe the problem

The `ssr` middleware outputted by adapter-node is not a proper middleware because it never calls `next()`, making it impossible to compose it with other middlewares.

### Describe the proposed solution

I'll submit a PR that'll introduce a new build option, `handle404?: boolean`:
- when set to true (or undefined), `handler` will handle 404 errors and display a custom error page
- when set to false, `handler` will call the next middleware on 404

### Alternatives considered

Make `handler` a pure middleware, and handle 404 in index.js, after `handler`:

https://github.com/sveltejs/kit/blob/668b0aaa2b5626e8f610f395ac551f98eafbd10b/packages/adapter-node/src/index.js#L11-L16

It represents a significant amount of work but allows small performance improvements when using only `handler`.

### Importance

i cannot use SvelteKit without it

**Comment by Conduitry** [comment]
I feel like this could/should be done more flexibly by having the Node adapter look for something on the `response` object that the user could set in the `handle` hook whenever they want - and not just when the status code is 404. I'm not sure what that API ought to look like exactly.

**Comment by benmccann** [comment]
Why can't you just use `handle` to return your custom 404 page?

**Comment by Conduitry** [comment]
I can see how it would be nice for SvelteKit's middleware to be proper middleware and not to have it responsible for calling the code you want it to fall through to.

All I disagree with is having it solely look for HTTP 404 response codes when deciding whether to do so.

**Comment by GauBen** [comment]
> I feel like this could/should be done more flexibly by having the Node adapter look for something on the `response` object that the user could set in the `handle` hook whenever they want - and not just when the status code is 404. I'm not sure what that API ought to look like exactly.

This would feel weird because the `handle` hook is made to be the end of the middleware chain

> Why can't you just use `handle` to return your custom 404 page?

I want to chain middlewares, not just handle 404 pages.
The idea is to chain of several SvelteKit apps, as can be seen here:
https://github.com/GauBen/modularized-sveltekit-app/blob/cf1f5dbb9261690a4f774f04a08959fa18d86d5f/common/index.js#L5-L14

**Comment by benmccann** [comment]
I wonder if it might be cleaner and better performance to instead pass a parameter to `app.render` that says whether to render 404 pages. Then `render` can return nothing instead of rendering a 404 page that you're just throwing away

**Comment by GauBen** [comment]
Yep I thought about this too so I browsed the code to see how hard it would be to implement...

I think about splitting `render` in two, making this part a stand-alone function:
https://github.com/sveltejs/kit/blob/c92f986277d12a8d6b74d94d20ed59b48f90bbdc/packages/kit/src/runtime/server/index.js#L169-L182

This would help create two middlewares:
- `handler`, a proper middlewares that invokes the next middleware on 404
- `notFoundHandler`, a final middleware that answers every request with a custom 404 pages (with `respond_with_error`)

If we all agree with this design, I might start working on a PR

**Comment by benmccann** [comment]
That sounds fairly reasonable for `adapter-node` at least. I haven't fully thought it through though. I think the main thing to pay attention to is making sure it works for all adapters. I'm not sure what your suggestion would look like in the other adapters

**Comment by Conduitry** [comment]
I still feel like this should be purely within the Node adapter. For most adapter outputs, being a passthrough middleware doesn't really make much sense to me, nor is it clear what that would do to the shared API.

**Comment by AlexRMU** [comment]
adapter-node generally needs to be improved. What if I want to limit the api? Install my AntiDDoS, metrics, cache, compress that would work outside the application? Use middlewares that use req/res from http? What if I want api requests not to get into the Svelte router at all?

Maybe create a new issue?

**Comment by Conduitry** [comment]
All of those things sound like they could be done with https://github.com/sveltejs/kit/tree/master/packages/adapter-node#custom-server
